[kotlin] support selection of datelibrary

magiepooh opened the pull request:
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

support selection of date library. (`java8`, `threetenbp`, `string`)

```bash
# no option
ags="generate -t modules/swagger-codegen/src/main/resources/kotlin-client -i modules/swagger-codegen/src/test/resources/2_0/petstore.yaml -l kotlin --artifact-id kotlin-petstore-client -D -o samples/client/petstore/kotlin $@"

# java8
ags="generate -t modules/swagger-codegen/src/main/resources/kotlin-client -i modules/swagger-codegen/src/test/resources/2_0/petstore.yaml -l kotlin --artifact-id kotlin-petstore-client -D dateLibrary=java8 -o samples/client/petstore/kotlin $@"

# no option
ags="generate -t modules/swagger-codegen/src/main/resources/kotlin-client -i modules/swagger-codegen/src/test/resources/2_0/petstore.yaml -l kotlin --artifact-id kotlin-petstore-client -D dateLibrary=string -o samples/client/petstore/kotlin $@"

# no option
ags="generate -t modules/swagger-codegen/src/main/resources/kotlin-client -i modules/swagger-codegen/src/test/resources/2_0/petstore.yaml -l kotlin --artifact-id kotlin-petstore-client -D dateLibrary=threetenbp -o samples/client/petstore/kotlin $@"
```

related Issue: https://github.com/swagger-api/swagger-codegen/issues/6892
(details of the change, additional tests that have been done, reference to the issue for tracking, etc)

wing328 commented:
@magiepooh thanks for the PR 👍 

cc @jimschubert @cbornet

jimschubert commented on modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/KotlinClientCodegen.java:
```diff
@@ -264,6 +291,26 @@ public void processOpts() {
+            typeMapping.put("Date", "kotlin.String");
+            typeMapping.put("DateTime", "kotlin.String");
+        } else if (DateLibrary.JAVA8.value.equals(dateLibrary)) {
+            additionalProperties.put(DateLibrary.JAVA8.value, "true");
```
`additionalProperties` should be `true` here rather than `"true"` if it's intended to be used for switching in templates.

jimschubert commented on modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/KotlinClientCodegen.java:
```diff
@@ -264,6 +291,26 @@ public void processOpts() {
+        }
+
+        if (DateLibrary.THREETENBP.value.equals(dateLibrary)) {
+            additionalProperties.put(DateLibrary.THREETENBP.value, "true");
```
additionalProperties should be true here rather than "true" if it's intended to be used for switching in templates.

jimschubert commented on modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/KotlinClientCodegen.java:
```diff
@@ -264,6 +291,26 @@ public void processOpts() {
+            importMapping.put("LocalDate", "org.threeten.bp.LocalDate");
+            importMapping.put("LocalDateTime", "org.threeten.bp.LocalDateTime");
+        } else if (DateLibrary.STRING.value.equals(dateLibrary)) {
+            additionalProperties.put(DateLibrary.STRING.value, "true");
```
I think having just `string` as an additional property would be confusing.

For instance, in a template this would lead to:

```
{{#string}}string date logic{{/string}}{{^string}}other logic{{/string}}
```

This seems like an unlikely use case, until it starts getting used and we can't change it without a breaking change. Also, I think this may cause confusion on some property/params which may have `string` properties as well.

additionalProperties should be true here rather than "true" if it's intended to be used for switching in templates.

Any objection to using `isStringDate` or something? iirc, Mustache will consider the string "true" for properties prefixed with `is` as a boolean. I could be wrong, but I prefer to set booleans explicitly. At one point, we were able to set them as strings but I don't know if that code was changed or if the Mustache dependency was updated but as of a few weeks ago they had to be explicit booleans.

jimschubert commented:
@magiepooh I don't know if you saw my comments? If you'd like, I can make the changes and open a PR against your fork.

tfujita-folio commented:
Thank you for reviewing! (I was not aware.)

> additionalProperties should be true here rather than "true" if it's intended to be used for switching in templates.

I'll fix it.

> https://github.com/swagger-api/swagger-codegen/pull/7054#discussion_r153106067

I think that selection regarding dates should be managed with a unified option.

`isStringDate` option and `dateLibrary` option may confuse the user. What do you think?

magiepooh commented:
Thank you for reviewing! (I was not aware.)

> additionalProperties should be true here rather than "true" if it's intended to be used for switching in templates.

I'll fix it.

> #7054 (comment)

I think that selection regarding dates should be managed with a unified option.

Multiple option (isStringDate option and dateLibrary option ) may confuse the user. What do you think?

jimschubert commented:
@maggiepooh I thought about it a little more. There's a `isString` and `isDate` on vars and params. Which of these would be true in this situation? I think it would be equally as confusing (maybe moreso) to have isString=false, isDate=true, and also have `string=true` available from additional properties. I think whatever is chosen will be confusing to consumers, mostly because `String` isn't a library for handling dates.

Maybe a better approach would be to set a private flag if the additional property is set (and remove it from additional properties to avoid any confusion), and post process variables and parameters to change from dates to strings. After all, that's what this option intends to do, correct? This way, it wouldn't interfere with templates that have `isDate` blocks.

I haven't looked at other generators to see if/how they handle this. @wing328 is there a common approach to forcing date/date-time to strings?

wing328 commented:
For PHP, we use our own serializer to handle the datetime to string conversion. For Java, we rely on 3rd-party libraries.

For Kotlin, can we follow Java's approach? (just throwing a suggestion but I've not looked into the Kotlin client code)

magiepooh commented:
> I thought about it a little more. There's a isString and isDate on vars and params. Which of these would be true in this situation? I think it would be equally as confusing (maybe moreso) to have isString=false, isDate=true, and also have string=true available from additional properties. I think whatever is chosen will be confusing to consumers, mostly because String isn't a library for handling dates.

I agree. 

@jimschubert Should I fix it as follows?
`additionalProperties.put(DateLibrary.STRING.value, "true");`
↓
`additionalProperties.put("isStringDate", "true");`

magiepooh commented:
I added additionalProperties for retrofit (HTTP client for Android).
However, now I thought that the string property is not necessary when creating a Retrofit adapter. So, I removed the string property.

wing328 commented:
@magiepooh CircleCI reports the following issue:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileTestKotlin'.
> Could not resolve all dependencies for configuration ':testCompile'.
   > Could not resolve org.hamcrest:hamcrest-core:1.3.
     Required by:
         io.swagger:kotlin-petstore-client:1.0.0 > io.kotlintest:kotlintest:2.0.2 > com.sksamuel.koors:koors:0.90.0 > junit:junit:4.12
      > Could not HEAD 'http://repo1.maven.org/maven2/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.pom'.
         > Connection reset

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED
```

I wonder if you can take a look when you've time.

magiepooh commented:
@wing328 

Can you help me? I don't know why CI failure.
My change is only `additionalProperties`. ( https://github.com/swagger-api/swagger-codegen/compare/f39dc67...8d0f661 )

wing328 commented:
I did a test with the option `threetenbp` but looks like there are some issues:
- build.gradle does not have the following to import the module:
```
compile group: 'org.threeten', name: 'threetenbp', version: '1.3.6'
```
- the following import statement is incorrect (in Order.kt)
```
import io.swagger.client.models.org.threeten.bp.LocalDateTime
```

I think I've solutions to these problems and will squeeze some time this weekend to work on it.

boylenssen commented:
Why is legacy not supported here? For Java, this sets it to `Date`. I would like this option for Kotlin too.

I used the snapshot of 2.4.0 (feb 10) and threetenbp works fine there. Legacy gives me Java8 datetime
